File: app/src/main/java/com/app/missednotificationsreminder/data/model/util/ApplicationIconHandler.java
package com.app.missednotificationsreminder.data.model.util;

import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;

import com.squareup.picasso.Picasso;
import com.squareup.picasso.Request;
import com.squareup.picasso.RequestHandler;

import java.io.IOException;

import timber.log.Timber;

/**
 * The application icon picasso handler
 */
public class ApplicationIconHandler extends RequestHandler {
    public static final String SCHEME = "appicon";

    private final PackageManager mPackageManager;

    public ApplicationIconHandler(PackageManager packageManager) {
        mPackageManager = packageManager;
    }

    @Override
    public boolean canHandleRequest(Request data) {
        return SCHEME.equals(data.uri.getScheme());
    }

    @Override
    public Result load(Request request, int networkPolicy) throws IOException {
        String packageName = request.uri.getHost();
        return new Result(getAppIcon(packageName), Picasso.LoadedFrom.DISK);
    }

    public Bitmap getAppIcon(String packageName) {

        try {
            Drawable drawable = mPackageManager.getApplicationIcon(packageName);

            if (drawable instanceof BitmapDrawable) {
                return ((BitmapDrawable) drawable).getBitmap();
            } else if (drawable != null) {
                final Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
                final Canvas canvas = new Canvas(bitmap);
                drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
                drawable.draw(canvas);
                return bitmap;
            } else {
                return null;
            }
        } catch (PackageManager.NameNotFoundException e) {
            Timber.e(e);
        }

        return null;
    }
}

File: app/src/main/java/com/app/missednotificationsreminder/settings/applicationselection/data/model/util/ApplicationIconHandler.kt
package com.app.missednotificationsreminder.settings.applicationselection.data.model.util

import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Canvas
import com.squareup.picasso.Picasso
import com.squareup.picasso.Request
import com.squareup.picasso.RequestHandler
import timber.log.Timber

/**
 * The application icon picasso handler
 */
class ApplicationIconHandler(private val mPackageManager: PackageManager) : RequestHandler() {
    override fun canHandleRequest(data: Request): Boolean {
        return SCHEME == data.uri.scheme
    }

    override fun load(request: Request, networkPolicy: Int): Result? {
        return getAppIcon(request.uri.host)?.let {
            Result(it, Picasso.LoadedFrom.DISK)
        }
    }

    private fun getAppIcon(packageName: String?): Bitmap? {
        try {
            return mPackageManager.getApplicationIcon(packageName)?.let { drawable ->
                val bitmap = Bitmap.createBitmap(drawable.intrinsicWidth, drawable.intrinsicHeight, Bitmap.Config.ARGB_8888)
                val canvas = Canvas(bitmap)
                drawable.setBounds(0, 0, canvas.width, canvas.height)
                drawable.draw(canvas)
                bitmap
            }
        } catch (e: Throwable) {
            Timber.e(e)
        }
        return null
    }

    companion object {
        const val SCHEME = "appicon"
    }

}